packages/template/-private/keywords/component.d.ts

```diff
@@ -1,19 +1,58 @@
-import { ComponentReturn, AnyBlocks, DirectInvokable, InvokableInstance } from '../integration';
+import { WithBoundArgs } from '../index';
+import {
+  ComponentReturn,
+  DirectInvokable,
+  InvokableInstance,
+  NamedArgs,
+  Invokable,
+  NamedArgNames,
+  AnyFunction,
+  UnwrapNamedArgs,
+} from '../integration';
+
+type PartiallyAppliedComponent<Component, Args> = Component extends Invokable<AnyFunction>
+  ? WithBoundArgs<
+      Component,
+      Exclude<Args extends NamedArgNames<Component> ? Args : never, typeof NamedArgs>
+    >
+  : never;
 
 export type ComponentKeyword = DirectInvokable<{
```

Comment:
This definition form hurts my head. 😂 

What would you think of extracting the type as an `interface` or `type` alias here? It took me actually opening it locally and following the definition of the `DirectInvokable` type to see that it was a function to actually be able to *parse* this properly as a type with two overloads of a function definition. Extracting it to a (non-exported) `type` alias should preserve the exact same semantics we have as is, but make it a bit clearer what's actually going on.

Reply:
(Also, I recognize that this definition form isn't new, but while it was a bit confusing before the overload *really* hurt my head when reading it now.)

Reply:
Pulling out an `interface` or `type` here would only remove the surrounding `DirectInvokable<...>` wrapper, I think. Beyond that the syntax would still be the same.

The only way to specify a multi-signature function type otherwise, to my knowledge, would be to declare the function as a value and then `typeof` it, which I'd lean away from—I don't love declaring nonexistent constants when it's avoidable.

Reply:
Yeah, I [eventually concluded](https://github.com/typed-ember/glint/pull/447#discussion_r1014509850) the same.